include/arch/x86/processor.h

```diff
@@ -241,7 +241,21 @@ typedef uint64_t x86_reg_t;
 typedef uint32_t x86_reg_t;
 #endif
 
-struct cpu_regs {
+struct __packed cpu_exc {
+    x86_ex_error_code_t error_code;
+    /* Populated by exception entry */
+    uint32_t vector;
```

Comment:
As you're touching this code, can we please clarify the above "wrong" order? The CPU pushes an error code for certain exceptions on its own (#PF, #GP, #DF,...), so one might expect a different ordering but as these two members are 32 bit and the error code is handled as a 64 bit value, it's actually fine -- we simply make use of the upper 32 bits of the hw/sw generated error code and stuff in the vector number.

I'd suggest an anonymous union like this, maybe?:

```c
union {
    struct {
        x86_ex_error_code_t error_code;
        /* Populated by exception entry */
        uint32_t vector;
    };
    /* We stuff the vector number into the upper 32 bit of the error code */
    x86_reg_t _raw;
};
```

No need to mention this all break apart under 32 bit but it's already broken, so no need to worry ;)

Reply:
This is handled by the exception_handler macro and the exception entry routines (entry_??) it generates.

The macro defines the entry_?? routines for all the exceptions like this:

```
exception_handler DE  X86_EX_DE  0
exception_handler DB  X86_EX_DB  0
exception_handler NMI X86_EX_NMI 0
exception_handler BP  X86_EX_BP  0
exception_handler OF  X86_EX_OF  0
exception_handler BR  X86_EX_BR  0
exception_handler UD  X86_EX_UD  0
exception_handler NM  X86_EX_NM  0
exception_handler DF  X86_EX_DF  1
exception_handler CS  X86_EX_CS  0
exception_handler TS  X86_EX_TS  1
exception_handler NP  X86_EX_NP  1
exception_handler SS  X86_EX_SS  1
exception_handler GP  X86_EX_GP  1
exception_handler PF  X86_EX_PF  1
exception_handler SPV X86_EX_SPV 0
exception_handler MF  X86_EX_MF  0
exception_handler AC  X86_EX_AC  1
exception_handler MC  X86_EX_MC  0
exception_handler XM  X86_EX_XM  0
exception_handler VE  X86_EX_VE  0
exception_handler SE  X86_EX_SE  1
```

and the macro implementation:
```
.macro exception_handler sym vec has_error_code
ENTRY(entry_\sym)
    enter_from_usermode

    .if \has_error_code == 0
        push $0
    .endif

    movl $\vec, cpu_exc_vector(%_ASM_SP)
    jmp handle_exception
END_FUNC(entry_\sym)
.endm
````

First parameter is the last part of the entry_?? name (the ??).
Second parameter is the actual exception vector.
The third parameter is an indicator whether the exception pushes error_code or not.

If the exception does not push the error_code the entry routine does so anyway (it pushes 0 then) to have the stack layout unified.

But, yeah it was broken for 32bit mode.

I will fix it with the following change:

```
--- a/arch/x86/entry.S
+++ b/arch/x86/entry.S
@@ -68,7 +68,12 @@ ENTRY(entry_\sym)
         push $0
     .endif
 
+#if defined(__x86_64__)
     movl $\vec, cpu_exc_vector(%_ASM_SP)
+#else
+    push $\vec
+#endif
+
     jmp handle_exception
 END_FUNC(entry_\sym)
 .endm
@@ -119,11 +124,7 @@ ENTRY(handle_exception)
 
     RESTORE_ALL_REGS
 
-#if defined(__x86_64__)
     add $8, %_ASM_SP
-#else
-    add $4, %_ASM_SP
-#endif
 
     enter_to_usermode
     IRET
```

Reply:
Unfortunately, that's not sufficient, as `struct cpu_exc` just doesn't fit the 32 bit exception frame (excessive padding for the segment registers, still mixed up ordering for `error_code` and `vector`). But that shouldn't hold the current PR.